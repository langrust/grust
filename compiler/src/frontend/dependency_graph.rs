mod component;
mod expression;
mod file;
mod statement;
mod stream_expression;
