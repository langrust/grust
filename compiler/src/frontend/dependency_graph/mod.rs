use petgraph::graphmap::DiGraphMap;

use crate::common::label::Label;

mod contract;
mod expression;
mod file;
mod node;
mod stream_expression;
mod statement;

/// Add an egde to the graph.
///
/// If edge already exits then keep the edge with the lowest weight.
pub fn add_edge(
    graph: &mut DiGraphMap<usize, Label>,
    signal_id: usize,
    dependency_id: usize,
    label: Label,
) {
    let prev_label = graph.add_edge(signal_id, dependency_id, label.clone());
    match (prev_label, label) {
        (Some(Label::Weight(prev_weight)), Label::Weight(new_weight))
            if prev_weight < new_weight =>
        {
            graph.add_edge(signal_id, dependency_id, Label::Weight(prev_weight));
        }
        _ => (),
    }
}
