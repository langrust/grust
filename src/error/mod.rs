use codespan_reporting::diagnostic::{Diagnostic, Label};

use crate::ast::{
    location::Location,
    type_system::Type,
};

/// Compilation errors enumeration.
/// 
/// [Error] enumeration is used during the compilation to alert
/// the programmer of some errors in its LanGRust program.
/// 
/// # Example
/// ```rust
/// use grustine::error::Error;
/// use grustine::ast::location::Location;
/// 
/// let mut errors = vec![];
/// 
/// let name = String::from("unknown");
/// let location = Location::default();
/// 
/// let error = Error::UnknownElement { name, location };
/// errors.push(error);
/// ```
/// 
#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Error {
    /// encountering an unknown element
    UnknownElement {
        /// the unknow identifier
        name: String,
        /// the error location
        location: Location,
    },
    /// incompatible application
    IncompatibleInputType {
        /// given type as input
        given_type: Type,
        /// expected type as input
        expected_type: Type,
        /// the error location
        location: Location,
    },
}
impl Error {
    /// Transform the error into a diagnostic.
    /// 
    /// This makes it possible to use [codespan_reporting] to print
    /// pretty errors.
    /// 
    /// # Example
    /// ```rust
    /// use grustine::error::Error;
    /// use grustine::ast::location::Location;
    /// use codespan_reporting::{
    ///     files::SimpleFiles,
    ///     term::termcolor::{StandardStream, ColorChoice},
    ///     term,
    /// };
    /// 
    /// let mut errors: Vec<Error> = vec![];
    /// let mut files = SimpleFiles::new();
    /// 
    /// let file_id = files.add("file_test.gr", "a code without x...");
    /// let name = String::from("x");
    /// let location = Location {
    ///     file_id,
    ///     range: 0..0,
    /// };
    /// 
    /// let error = Error::UnknownElement { name, location };
    /// errors.push(error);
    /// 
    /// let writer = StandardStream::stderr(ColorChoice::Always);
    /// let config = term::Config::default();
    /// for error in &errors {
    ///     let writer = &mut writer.lock();
    ///     let _ = term::emit(writer, &config, &files, &error.to_diagnostic());
    /// }
    /// ```
    pub fn to_diagnostic(&self) -> Diagnostic<usize> {
        match self {
            Error::UnknownElement { name, location } => Diagnostic::error()
                .with_message("unknown element")
                .with_labels(vec![
                    Label::primary(location.file_id, location.range.clone())
                        .with_message("unknown")
                ])
                .with_notes(vec![
                    format!("element '{name}' is not defined")
                ]
            ),
            Error::IncompatibleInputType { given_type, expected_type, location } => Diagnostic::error()
                .with_message("incompatible application")
                .with_labels(vec![
                    Label::primary(location.file_id, location.range.clone())
                        .with_message("wrong input type")
                ])
                .with_notes(vec![
                    format!("expected '{expected_type}' but '{given_type}' was given")
                ]
            ),
        }
    }
}
