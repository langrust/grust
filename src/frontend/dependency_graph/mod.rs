use petgraph::graphmap::DiGraphMap;

use crate::common::label::Label;

mod expression;
mod file;
mod node;
mod stream_expression;
mod term;

pub fn add_edge(
    graph: &mut DiGraphMap<usize, Label>,
    signal_id: usize,
    dependency_id: usize,
    label: Label,
) {
    let prev_label = graph.add_edge(signal_id, dependency_id, label.clone());
    match (prev_label, label) {
        (Some(Label::Weight(prev_weight)), Label::Weight(new_weight))
            if prev_weight < new_weight =>
        {
            graph.add_edge(signal_id, dependency_id, Label::Weight(prev_weight));
        }
        _ => (),
    }
}
