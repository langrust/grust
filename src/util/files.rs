use codespan_reporting::files;
use std::ops::Range;

/// A file that is backed by an `Arc<String>`.
#[derive(Debug, Clone)]
struct File {
    /// The name of the file.
    name: String,
    /// The source code of the file.
    source: String,
    /// The starting byte indices in the source code.
    line_starts: Vec<usize>,
}

impl File {
    fn line_start(&self, line_index: usize) -> Result<usize, files::Error> {
        use std::cmp::Ordering;

        match line_index.cmp(&self.line_starts.len()) {
            Ordering::Less => Ok(*self
                .line_starts
                .get(line_index)
                .expect("failed despite previous check")),
            Ordering::Equal => Ok(self.source.len()),
            Ordering::Greater => Err(files::Error::LineTooLarge {
                given: line_index,
                max: self.line_starts.len() - 1,
            }),
        }
    }
}

/// An opaque file identifier.
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub struct FileId(i32);
impl FileId {
    /// Construct and return the default file identifier.
    pub fn default() -> Self {
        return FileId(-1);
    }
}

/// The files handler.
#[derive(Debug, Clone)]
pub struct Files {
    files: Vec<File>,
}

impl Files {
    /// Create a new files database.
    pub fn new() -> Files {
        Files { files: Vec::new() }
    }

    /// Add a file to the database, returning the handle that can be used to
    /// refer to it again.
    pub fn add(
        &mut self,
        name: impl Into<String>,
        source: impl Into<String>,
    ) -> Option<FileId> {
        let file_id = FileId(i32::try_from(self.files.len()).ok()?);
        let name = name.into();
        let source = source.into();
        let line_starts = files::line_starts(&source).collect();

        self.files.push(File {
            name,
            line_starts,
            source,
        });

        Some(file_id)
    }

    /// Get the file corresponding to the given id.
    fn get(&self, file_id: FileId) -> Result<&File, files::Error> {
        self.files
            .get(file_id.0 as usize)
            .ok_or(files::Error::FileMissing)
    }
}

impl<'files> files::Files<'files> for Files {
    type FileId = FileId;
    type Name = &'files str;
    type Source = &'files str;

    fn name(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(self.get(file_id)?.name.as_ref())
    }

    fn source(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(&self.get(file_id)?.source)
    }

    fn line_index(
        &self,
        file_id: FileId,
        byte_index: usize
    ) -> Result<usize, files::Error> {
        self.get(file_id)?
            .line_starts
            .binary_search(&byte_index)
            .or_else(|next_line| Ok(next_line - 1))
    }

    fn line_range(
        &self,
        file_id: FileId,
        line_index: usize,
    ) -> Result<Range<usize>, files::Error> {
        let file = self.get(file_id)?;
        let line_start = file.line_start(line_index)?;
        let next_line_start = file.line_start(line_index + 1)?;

        Ok(line_start..next_line_start)
    }
}