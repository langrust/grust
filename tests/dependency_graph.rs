use codespan_reporting::files::SimpleFiles;

use grustine::dependency_graph;
use grustine::error::display;

#[test]
fn dependency_graph_of_counter() {
    let mut files = SimpleFiles::new();

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/dependency_graph/success/counter.gr")
            .expect("unkown file"),
    );

    match dependency_graph(counter_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_blinking() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/dependency_graph/success/blinking.gr")
            .expect("unkown file"),
    );

    match dependency_graph(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_button_management() {
    let mut files = SimpleFiles::new();

    let button_management_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/dependency_graph/success/button_management.gr")
            .expect("unkown file"),
    );

    match dependency_graph(button_management_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_button_management_condition_match() {
    let mut files = SimpleFiles::new();

    let button_management_condition_match_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string(
            "tests/fixture/dependency_graph/success/button_management_condition_match.gr",
        )
        .expect("unkown file"),
    );

    match dependency_graph(button_management_condition_match_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_button_management_using_function() {
    let mut files = SimpleFiles::new();

    let button_management_using_function_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string(
            "tests/fixture/dependency_graph/success/button_management_using_function.gr",
        )
        .expect("unkown file"),
    );

    match dependency_graph(button_management_using_function_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_pid() {
    let mut files = SimpleFiles::new();

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/dependency_graph/success/pid.gr")
            .expect("unkown file"),
    );

    match dependency_graph(pid_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}
