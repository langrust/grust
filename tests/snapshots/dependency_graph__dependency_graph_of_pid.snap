---
source: tests/dependency_graph.rs
expression: file
---
typedefs:
  - Structure:
      id: GainPID
      fields:
        - - k_p
          - Float
        - - k_i
          - Float
        - - k_d
          - Float
      location:
        file_id: 0
        range:
          start: 0
          end: 66
functions:
  - id: min
    inputs:
      - - a
        - Float
      - - b
        - Float
    statements:
      - id: test
        element_type: Boolean
        expression:
          Application:
            function_expression:
              Call:
                id: " > "
                typing:
                  Abstract:
                    - - Float
                      - Float
                    - Boolean
                location:
                  file_id: 0
                  range:
                    start: 304
                    end: 305
            inputs:
              - Call:
                  id: a
                  typing: Float
                  location:
                    file_id: 0
                    range:
                      start: 302
                      end: 303
              - Call:
                  id: b
                  typing: Float
                  location:
                    file_id: 0
                    range:
                      start: 306
                      end: 307
            typing: Boolean
            location:
              file_id: 0
              range:
                start: 302
                end: 307
        location:
          file_id: 0
          range:
            start: 285
            end: 308
      - id: result
        element_type: Float
        expression:
          Application:
            function_expression:
              Call:
                id: if_then_else
                typing:
                  Abstract:
                    - - Boolean
                      - Float
                      - Float
                    - Float
                location:
                  file_id: 0
                  range:
                    start: 333
                    end: 354
            inputs:
              - Call:
                  id: test
                  typing: Boolean
                  location:
                    file_id: 0
                    range:
                      start: 336
                      end: 340
              - Call:
                  id: b
                  typing: Float
                  location:
                    file_id: 0
                    range:
                      start: 346
                      end: 347
              - Call:
                  id: a
                  typing: Float
                  location:
                    file_id: 0
                    range:
                      start: 353
                      end: 354
            typing: Float
            location:
              file_id: 0
              range:
                start: 333
                end: 354
        location:
          file_id: 0
          range:
            start: 313
            end: 355
    returned:
      - Float
      - Call:
          id: result
          typing: Float
          location:
            file_id: 0
            range:
              start: 367
              end: 373
    location:
      file_id: 0
      range:
        start: 237
        end: 376
nodes:
  - id: integrate
    is_component: false
    inputs:
      - - x
        - Float
      - - dt
        - Float
    unscheduled_equations:
      MAX_INTEGRALE:
        scope: Local
        id: MAX_INTEGRALE
        signal_type: Float
        expression:
          Constant:
            constant:
              Float: 1000
            typing: Float
            location:
              file_id: 0
              range:
                start: 226
                end: 232
            dependencies: []
        location:
          file_id: 0
          range:
            start: 203
            end: 233
      i:
        scope: Output
        id: i
        signal_type: Float
        expression:
          MapApplication:
            function_expression:
              Call:
                id: min
                typing:
                  Abstract:
                    - - Float
                      - Float
                    - Float
                location:
                  file_id: 0
                  range:
                    start: 193
                    end: 196
            inputs:
              - MapApplication:
                  function_expression:
                    Call:
                      id: " + "
                      typing:
                        Abstract:
                          - - Float
                            - Float
                          - Float
                      location:
                        file_id: 0
                        range:
                          start: 164
                          end: 165
                  inputs:
                    - SignalCall:
                        signal:
                          id: prev_i
                          scope: Local
                        typing: Float
                        location:
                          file_id: 0
                          range:
                            start: 157
                            end: 163
                        dependencies:
                          - - prev_i
                            - 0
                    - MapApplication:
                        function_expression:
                          Call:
                            id: " * "
                            typing:
                              Abstract:
                                - - Float
                                  - Float
                                - Float
                            location:
                              file_id: 0
                              range:
                                start: 168
                                end: 169
                        inputs:
                          - SignalCall:
                              signal:
                                id: x
                                scope: Input
                              typing: Float
                              location:
                                file_id: 0
                                range:
                                  start: 166
                                  end: 167
                              dependencies:
                                - - x
                                  - 0
                          - SignalCall:
                              signal:
                                id: dt
                                scope: Input
                              typing: Float
                              location:
                                file_id: 0
                                range:
                                  start: 170
                                  end: 172
                              dependencies:
                                - - dt
                                  - 0
                        typing: Float
                        location:
                          file_id: 0
                          range:
                            start: 166
                            end: 172
                        dependencies:
                          - - x
                            - 0
                          - - dt
                            - 0
                  typing: Float
                  location:
                    file_id: 0
                    range:
                      start: 157
                      end: 172
                  dependencies:
                    - - prev_i
                      - 0
                    - - x
                      - 0
                    - - dt
                      - 0
              - SignalCall:
                  signal:
                    id: MAX_INTEGRALE
                    scope: Local
                  typing: Float
                  location:
                    file_id: 0
                    range:
                      start: 174
                      end: 187
                  dependencies:
                    - - MAX_INTEGRALE
                      - 0
            typing: Float
            location:
              file_id: 0
              range:
                start: 156
                end: 197
            dependencies:
              - - prev_i
                - 0
              - - x
                - 0
              - - dt
                - 0
              - - MAX_INTEGRALE
                - 0
        location:
          file_id: 0
          range:
            start: 141
            end: 198
      prev_i:
        scope: Local
        id: prev_i
        signal_type: Float
        expression:
          FollowedBy:
            constant:
              Float: 0
            expression:
              SignalCall:
                signal:
                  id: i
                  scope: Output
                typing: Float
                location:
                  file_id: 0
                  range:
                    start: 134
                    end: 135
                dependencies:
                  - - i
                    - 0
            typing: Float
            location:
              file_id: 0
              range:
                start: 126
                end: 135
            dependencies:
              - - i
                - 1
        location:
          file_id: 0
          range:
            start: 110
            end: 136
    unitary_nodes: {}
    location:
      file_id: 0
      range:
        start: 68
        end: 235
    graph:
      vertices:
        MAX_INTEGRALE:
          id: MAX_INTEGRALE
          value: Black
          neighbors: []
        dt:
          id: dt
          value: Black
          neighbors: []
        i:
          id: i
          value: Black
          neighbors:
            - id: prev_i
              weight: 0
            - id: x
              weight: 0
            - id: dt
              weight: 0
            - id: MAX_INTEGRALE
              weight: 0
        prev_i:
          id: prev_i
          value: Black
          neighbors:
            - id: i
              weight: 1
        x:
          id: x
          value: Black
          neighbors: []
  - id: derive
    is_component: false
    inputs:
      - - x
        - Float
      - - dt
        - Float
    unscheduled_equations:
      d:
        scope: Output
        id: d
        signal_type: Float
        expression:
          MapApplication:
            function_expression:
              Call:
                id: " / "
                typing:
                  Abstract:
                    - - Float
                      - Float
                    - Float
                location:
                  file_id: 0
                  range:
                    start: 444
                    end: 445
            inputs:
              - MapApplication:
                  function_expression:
                    Call:
                      id: (_)
                      typing:
                        Abstract:
                          - - Float
                          - Float
                      location:
                        file_id: 0
                        range:
                          start: 432
                          end: 444
                  inputs:
                    - MapApplication:
                        function_expression:
                          Call:
                            id: " - "
                            typing:
                              Abstract:
                                - - Float
                                  - Float
                                - Float
                            location:
                              file_id: 0
                              range:
                                start: 435
                                end: 436
                        inputs:
                          - SignalCall:
                              signal:
                                id: x
                                scope: Input
                              typing: Float
                              location:
                                file_id: 0
                                range:
                                  start: 433
                                  end: 434
                              dependencies:
                                - - x
                                  - 0
                          - SignalCall:
                              signal:
                                id: prev_x
                                scope: Local
                              typing: Float
                              location:
                                file_id: 0
                                range:
                                  start: 437
                                  end: 443
                              dependencies:
                                - - prev_x
                                  - 0
                        typing: Float
                        location:
                          file_id: 0
                          range:
                            start: 433
                            end: 443
                        dependencies:
                          - - x
                            - 0
                          - - prev_x
                            - 0
                  typing: Float
                  location:
                    file_id: 0
                    range:
                      start: 432
                      end: 444
                  dependencies:
                    - - x
                      - 0
                    - - prev_x
                      - 0
              - SignalCall:
                  signal:
                    id: dt
                    scope: Input
                  typing: Float
                  location:
                    file_id: 0
                    range:
                      start: 445
                      end: 447
                  dependencies:
                    - - dt
                      - 0
            typing: Float
            location:
              file_id: 0
              range:
                start: 432
                end: 447
            dependencies:
              - - x
                - 0
              - - prev_x
                - 0
              - - dt
                - 0
        location:
          file_id: 0
          range:
            start: 417
            end: 448
      prev_x:
        scope: Local
        id: prev_x
        signal_type: Float
        expression:
          FollowedBy:
            constant:
              Float: 0
            expression:
              SignalCall:
                signal:
                  id: x
                  scope: Input
                typing: Float
                location:
                  file_id: 0
                  range:
                    start: 477
                    end: 478
                dependencies:
                  - - x
                    - 0
            typing: Float
            location:
              file_id: 0
              range:
                start: 469
                end: 478
            dependencies:
              - - x
                - 1
        location:
          file_id: 0
          range:
            start: 453
            end: 479
    unitary_nodes: {}
    location:
      file_id: 0
      range:
        start: 378
        end: 481
    graph:
      vertices:
        d:
          id: d
          value: Black
          neighbors:
            - id: x
              weight: 0
            - id: prev_x
              weight: 0
            - id: dt
              weight: 0
        dt:
          id: dt
          value: Black
          neighbors: []
        prev_x:
          id: prev_x
          value: Black
          neighbors:
            - id: x
              weight: 1
        x:
          id: x
          value: Black
          neighbors: []
component:
  id: pid
  is_component: true
  inputs:
    - - v_c
      - Float
    - - v
      - Float
    - - dt
      - Float
  unscheduled_equations:
    e:
      scope: Local
      id: e
      signal_type: Float
      expression:
        MapApplication:
          function_expression:
            Call:
              id: " - "
              typing:
                Abstract:
                  - - Float
                    - Float
                  - Float
              location:
                file_id: 0
                range:
                  start: 768
                  end: 769
          inputs:
            - SignalCall:
                signal:
                  id: v_c
                  scope: Input
                typing: Float
                location:
                  file_id: 0
                  range:
                    start: 764
                    end: 767
                dependencies:
                  - - v_c
                    - 0
            - SignalCall:
                signal:
                  id: v
                  scope: Input
                typing: Float
                location:
                  file_id: 0
                  range:
                    start: 770
                    end: 771
                dependencies:
                  - - v
                    - 0
          typing: Float
          location:
            file_id: 0
            range:
              start: 764
              end: 771
          dependencies:
            - - v_c
              - 0
            - - v
              - 0
      location:
        file_id: 0
        range:
          start: 753
          end: 772
    e_d:
      scope: Local
      id: e_d
      signal_type: Float
      expression:
        NodeApplication:
          node: derive
          inputs:
            - SignalCall:
                signal:
                  id: e
                  scope: Local
                typing: Float
                location:
                  file_id: 0
                  range:
                    start: 834
                    end: 835
                dependencies:
                  - - e
                    - 0
            - SignalCall:
                signal:
                  id: dt
                  scope: Input
                typing: Float
                location:
                  file_id: 0
                  range:
                    start: 837
                    end: 839
                dependencies:
                  - - dt
                    - 0
          signal: d
          typing: Float
          location:
            file_id: 0
            range:
              start: 827
              end: 842
          dependencies:
            - - e
              - 0
            - - e
              - 1
            - - dt
              - 0
      location:
        file_id: 0
        range:
          start: 814
          end: 843
    e_i:
      scope: Local
      id: e_i
      signal_type: Float
      expression:
        NodeApplication:
          node: integrate
          inputs:
            - SignalCall:
                signal:
                  id: e
                  scope: Local
                typing: Float
                location:
                  file_id: 0
                  range:
                    start: 800
                    end: 801
                dependencies:
                  - - e
                    - 0
            - SignalCall:
                signal:
                  id: dt
                  scope: Input
                typing: Float
                location:
                  file_id: 0
                  range:
                    start: 803
                    end: 805
                dependencies:
                  - - dt
                    - 0
          signal: i
          typing: Float
          location:
            file_id: 0
            range:
              start: 790
              end: 808
          dependencies:
            - - e
              - 0
            - - dt
              - 0
      location:
        file_id: 0
        range:
          start: 777
          end: 809
    gain:
      scope: Local
      id: gain
      signal_type:
        Structure: GainPID
      expression:
        Structure:
          name: GainPID
          fields:
            - - k_p
              - Constant:
                  constant:
                    Float: 0.2
                  typing: Float
                  location:
                    file_id: 0
                    range:
                      start: 575
                      end: 578
                  dependencies: []
            - - k_i
              - Constant:
                  constant:
                    Float: 1.5
                  typing: Float
                  location:
                    file_id: 0
                    range:
                      start: 593
                      end: 596
                  dependencies: []
            - - k_d
              - Constant:
                  constant:
                    Float: 6
                  typing: Float
                  location:
                    file_id: 0
                    range:
                      start: 611
                      end: 614
                  dependencies: []
          typing:
            Structure: GainPID
          location:
            file_id: 0
            range:
              start: 552
              end: 621
          dependencies: []
      location:
        file_id: 0
        range:
          start: 536
          end: 622
    u:
      scope: Output
      id: u
      signal_type: Float
      expression:
        Match:
          expression:
            SignalCall:
              signal:
                id: gain
                scope: Local
              typing:
                Structure: GainPID
              location:
                file_id: 0
                range:
                  start: 649
                  end: 653
              dependencies:
                - - gain
                  - 0
          arms:
            - - Structure:
                  name: GainPID
                  fields:
                    - - k_p
                      - Identifier:
                          name: k_p
                          location:
                            file_id: 0
                            range:
                              start: 680
                              end: 683
                    - - k_i
                      - Identifier:
                          name: k_i
                          location:
                            file_id: 0
                            range:
                              start: 690
                              end: 693
                    - - k_d
                      - Identifier:
                          name: k_d
                          location:
                            file_id: 0
                            range:
                              start: 700
                              end: 703
                  location:
                    file_id: 0
                    range:
                      start: 665
                      end: 705
              - ~
              - []
              - MapApplication:
                  function_expression:
                    Call:
                      id: " + "
                      typing:
                        Abstract:
                          - - Float
                            - Float
                          - Float
                      location:
                        file_id: 0
                        range:
                          start: 729
                          end: 730
                  inputs:
                    - MapApplication:
                        function_expression:
                          Call:
                            id: " + "
                            typing:
                              Abstract:
                                - - Float
                                  - Float
                                - Float
                            location:
                              file_id: 0
                              range:
                                start: 717
                                end: 718
                        inputs:
                          - MapApplication:
                              function_expression:
                                Call:
                                  id: " * "
                                  typing:
                                    Abstract:
                                      - - Float
                                        - Float
                                      - Float
                                  location:
                                    file_id: 0
                                    range:
                                      start: 713
                                      end: 714
                              inputs:
                                - SignalCall:
                                    signal:
                                      id: k_p
                                      scope: Local
                                    typing: Float
                                    location:
                                      file_id: 0
                                      range:
                                        start: 709
                                        end: 712
                                    dependencies:
                                      - - k_p
                                        - 0
                                - SignalCall:
                                    signal:
                                      id: e
                                      scope: Local
                                    typing: Float
                                    location:
                                      file_id: 0
                                      range:
                                        start: 715
                                        end: 716
                                    dependencies:
                                      - - e
                                        - 0
                              typing: Float
                              location:
                                file_id: 0
                                range:
                                  start: 709
                                  end: 716
                              dependencies:
                                - - k_p
                                  - 0
                                - - e
                                  - 0
                          - MapApplication:
                              function_expression:
                                Call:
                                  id: " * "
                                  typing:
                                    Abstract:
                                      - - Float
                                        - Float
                                      - Float
                                  location:
                                    file_id: 0
                                    range:
                                      start: 723
                                      end: 724
                              inputs:
                                - SignalCall:
                                    signal:
                                      id: k_i
                                      scope: Local
                                    typing: Float
                                    location:
                                      file_id: 0
                                      range:
                                        start: 719
                                        end: 722
                                    dependencies:
                                      - - k_i
                                        - 0
                                - SignalCall:
                                    signal:
                                      id: e_i
                                      scope: Local
                                    typing: Float
                                    location:
                                      file_id: 0
                                      range:
                                        start: 725
                                        end: 728
                                    dependencies:
                                      - - e_i
                                        - 0
                              typing: Float
                              location:
                                file_id: 0
                                range:
                                  start: 719
                                  end: 728
                              dependencies:
                                - - k_i
                                  - 0
                                - - e_i
                                  - 0
                        typing: Float
                        location:
                          file_id: 0
                          range:
                            start: 709
                            end: 728
                        dependencies:
                          - - k_p
                            - 0
                          - - e
                            - 0
                          - - k_i
                            - 0
                          - - e_i
                            - 0
                    - MapApplication:
                        function_expression:
                          Call:
                            id: " * "
                            typing:
                              Abstract:
                                - - Float
                                  - Float
                                - Float
                            location:
                              file_id: 0
                              range:
                                start: 735
                                end: 736
                        inputs:
                          - SignalCall:
                              signal:
                                id: k_d
                                scope: Local
                              typing: Float
                              location:
                                file_id: 0
                                range:
                                  start: 731
                                  end: 734
                              dependencies:
                                - - k_d
                                  - 0
                          - SignalCall:
                              signal:
                                id: e_d
                                scope: Local
                              typing: Float
                              location:
                                file_id: 0
                                range:
                                  start: 737
                                  end: 740
                              dependencies:
                                - - e_d
                                  - 0
                        typing: Float
                        location:
                          file_id: 0
                          range:
                            start: 731
                            end: 740
                        dependencies:
                          - - k_d
                            - 0
                          - - e_d
                            - 0
                  typing: Float
                  location:
                    file_id: 0
                    range:
                      start: 709
                      end: 740
                  dependencies:
                    - - k_p
                      - 0
                    - - e
                      - 0
                    - - k_i
                      - 0
                    - - e_i
                      - 0
                    - - k_d
                      - 0
                    - - e_d
                      - 0
          typing: Float
          location:
            file_id: 0
            range:
              start: 642
              end: 747
          dependencies:
            - - e
              - 0
            - - e_i
              - 0
            - - e_d
              - 0
            - - gain
              - 0
      location:
        file_id: 0
        range:
          start: 627
          end: 748
  unitary_nodes: {}
  location:
    file_id: 0
    range:
      start: 483
      end: 845
  graph:
    vertices:
      dt:
        id: dt
        value: Black
        neighbors: []
      e:
        id: e
        value: Black
        neighbors:
          - id: v_c
            weight: 0
          - id: v
            weight: 0
      e_d:
        id: e_d
        value: Black
        neighbors:
          - id: e
            weight: 0
          - id: e
            weight: 1
          - id: dt
            weight: 0
      e_i:
        id: e_i
        value: Black
        neighbors:
          - id: e
            weight: 0
          - id: dt
            weight: 0
      gain:
        id: gain
        value: Black
        neighbors: []
      u:
        id: u
        value: Black
        neighbors:
          - id: e
            weight: 0
          - id: e_i
            weight: 0
          - id: e_d
            weight: 0
          - id: gain
            weight: 0
      v:
        id: v
        value: Black
        neighbors: []
      v_c:
        id: v_c
        value: Black
        neighbors: []
location:
  file_id: 0
  range:
    start: 0
    end: 845

